bzlmod/private/toolchains/toolchains.bzl

```diff
@@ -0,0 +1,116 @@
+load("//crate_universe:deps_bootstrap.bzl", _cargo_bazel_bootstrap_repo_rule = "cargo_bazel_bootstrap")
+load("//rust/platform:triple.bzl", "get_host_triple")
+load("//rust/private:common.bzl", "rust_common")
+load(
+    "//rust/private:repository_utils.bzl",
+    "DEFAULT_EXTRA_TARGET_TRIPLES",
+    "DEFAULT_NIGHTLY_VERSION",
+    "DEFAULT_STATIC_RUST_URL_TEMPLATES",
+)
+load("//rust:repositories.bzl", "rust_register_toolchains", "rust_toolchain_tools_repository")
+
+def generate_uid(toolchain):
+    """
+    Generates a UID representing a toolchain. This allows us to share toolchains
+    between different modules if they happen to use the same configuration.
+    """
+    unique_config = [
+        toolchain.allocator_library,
+        str(toolchain.dev_components),
+        toolchain.edition,
+        toolchain.rust_analyzer_version,
+        toolchain.rustfmt_version,
+    ]
+    unique_config.append("sha256s")
+    for k, v in toolchain.sha256s:
+        unique_config.append(k)
+        unique_config.append(v)
+    unique_config.append("extra_target_triples")
+    unique_config.extend(toolchain.extra_target_triples)
+    unique_config.append("urls")
+    unique_config.extend(toolchain.urls)
+    unique_config.append("versions")
+    unique_config.extend(toolchain.versions)
+    return hash("".join([str(hash(i)) for i in unique_config]))
+
+def _toolchains_impl(module_ctx):
+    host_triple = get_host_triple(module_ctx)
+
+    toolchains = {}
+    for mod in module_ctx.modules:
+        for toolchain in mod.tags.host_tools:
+            rust_toolchain_tools_repository(
+                name = "%s_rust_host_tools" % mod.name,
+                exec_triple = host_triple.str,
+                target_triple = host_triple.str,
+                allocator_library = toolchain.allocator_library,
+                version = toolchain.version,
+                rustfmt_version = toolchain.rustfmt_version,
+                edition = toolchain.edition,
+                sha256s = toolchain.sha256s,
+                urls = toolchain.urls,
+                dev_components = False,
+                auth = None,
+            )
+
+        for toolchain in mod.tags.toolchain:
+            repo_name = "%s_rust_toolchains" % mod.name
```

Comment:
It seems like you're making it so that the whole module graph can create and register its own toolchains, but isn't the set of available toolchains a global thing? That's why I originally made it so that only the top-level module can define toolchains, but maybe I'm misunderstanding something here.

Reply:
To be honest, I'm not particularly familiar with toolchains, so you could definitely be right. I was making the assumption that each module registered its own toolchains, and that, like use_repo, each module had its own visibility into the set of toolchains, and specifically used  ones that they defined themselves?

Eg. I saw that your toolchain definition has the tag "edition". Since editions are specifically not backwards compatible, something designed for a 2018 edition might fail if compiled with a toolchain that specifies the 2021 edition.

@Wyverald Could you confirm how register_toolchain works? Is it global or per-module?